org-code-javabuilder/app/src/main/java/org/code/javabuilder/JavaBuildController.java

```diff
@@ -21,9 +38,122 @@
+      userProgram.setFileName(filename.substring(0, filename.indexOf(".java")));
+    }
+    executeHelper(userProgram, principal);
+    return new UserProgramOutput("Done!");
+  }
+
+  public void executeHelper(UserProgram userProgram, Principal principal) {
+    File tempFolder = null;
+    try {
+      tempFolder = Files.createTempDirectory("tmpdir").toFile();
+      DiagnosticCollector<JavaFileObject> diagnostics = new DiagnosticCollector<JavaFileObject>();
+
+      CompilationTask task = getCompilationTask(userProgram, tempFolder, diagnostics);
+      compileRunService.sendMessages(principal.getName(), "Compiling your program...");
+      boolean success = task.call();
+
+      // diagnostics will include any compiler errors
+      for (Diagnostic<? extends JavaFileObject> diagnostic : diagnostics.getDiagnostics()) {
+        compileRunService.sendMessages(principal.getName(), diagnostic.toString());
+      }
+
+      if (success) {
+        // set System.out to be a specific output stream in order to capture output of the
+        // program and send it back to the user
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        PrintStream out = new PrintStream(outputStream);
+        System.setOut(out);
+
+        compileRunService.sendMessages(principal.getName(), "Compiled!");
+        compileRunService.sendMessages(principal.getName(), "Running your program...");
+        runClass(tempFolder.toURI().toURL(), userProgram.getFileName(), principal);
+
+        // outputStream should now contain output of userProgram
+        outputStream.flush();
+        String result = outputStream.toString();
+        if (result.length() > 0) {
+          compileRunService.sendMessages(principal.getName(), result);
+        }
+      } else {
+        compileRunService.sendMessages(
+            principal.getName(), "There was an error compiling your program.");
+      }
+
+    } catch (IOException e) {
+      // IOException could be called by creating a temporary folder or writing to that folder.
+      // May need better error handling for this.
+      compileRunService.sendMessages(
+          principal.getName(), "There was an issue trying to run your program, please try again.");
+      System.out.println(e.getStackTrace());
+    }
+
+    if (tempFolder != null) {
+      tempFolder.delete();
+    }
+
+    // ensure System.out is reset
+    System.setOut(System.out);
+  }
+
+  // Given a user program, create a compilation task that will save the .class file to the given
+  // temp folder and output any compilation messages to diagnostics.
+  public CompilationTask getCompilationTask(
+      UserProgram userProgram, File tempFolder, DiagnosticCollector<JavaFileObject> diagnostics)
+      throws IOException {
+    JavaCompiler compiler = ToolProvider.getSystemJavaCompiler();
+
+    // set output of compilation to be a temporary folder
+    StandardJavaFileManager fileManager = compiler.getStandardFileManager(null, null, null);
```

Comment:
FWIW, IntelliJ will do this kind of annotation automatically when reading code in the IDE.

Reply:
A lot of these Java library functions have a bunch of parameters that are often unnecessary, so I am in favor of not over-commenting and relying on IDE capabilities to inform us of what the parameters are. It's also harder to mess up parameters in Java than in a language like Javascript or Ruby because everything is typed (this is a bad example because everything is null).

Reply:
👍